clingraph/__init__.py

```diff
@@ -1,188 +1,305 @@
+                g = graphs[vals['graph']]
+                default_used = default_used or vals['graph'] == self.default_graph
+                if e == 'node':
+                    g.node(e_id, **vals['attr'])
+                else:
+                    g.edge(e_id[0], e_id[1], **vals['attr'])
+
+        if not default_used:
+            del structure['graph'][self.default_graph]
+            del graphs[self.default_graph]
+
+        return graphs
+
+    def model2graphs(self, model):
+        """
+        Construct the corresponing graphs from a model
+        Arguments:
+            model: Clingos stable model
+        """
+        structure = self.model2structure(model)
+
+        graphs = self.structure2graphs(structure, f'{model.number:04d}')
+
+        # Attach subgraphs
+        final_graphs = []
+        graph_order = [(g_id, g['graph'] or 'top')
+                       for g_id, g in structure['graph'].items()]
+        nx_g = nx.DiGraph(graph_order)
+        final_order = list(nx.topological_sort(nx_g))
+
+        for g_id in final_order:
+            if g_id == 'top':
+                continue
+
+            g_parent = structure['graph'][g_id]['graph']
+            g = graphs[g_id]
+            if g_parent is None:
+                final_graphs.append(g)
+                continue
+            g.name = 'cluster_'+g.name
+            graphs[g_parent].subgraph(g)
+
+        return final_graphs
+
+    def save(self):
+        """
+        Saves the graphs
+        """
+        for _, graphs in self.graphs.items():
+            for graph in graphs:
+                render_format = self.format in ['pdf', 'svg', 'png']
+                if render_format or self.view or self.gif:
+                    graph.render(
+                        view=self.view,
+                        cleanup=True,
+                        quiet_view=self.view,
+                    )
+                else:
+                    if not os.path.exists(graph.directory):
+                        os.makedirs(graph.directory)
+                    with open(graph.filepath, "w+") as f:
+                        f.write(graph.source)
+
+            if self.gif:
+                images = []
+                for graph in graphs:
+                    images.append(imageio.imread(
+                        graph.filepath+"."+graph.format))
+                imageio.mimsave(os.path.join(
+                    graph.directory, 'movie.gif'), images, fps=1)
+
+    def print_model(self, model, printer):
+        """
+        Function to print additional information when the text output is used.
+        Arguments:
+            model: Clingos stable model
+        """
+        print(model)
+        print("\n" + "-"*20 + ' Clingraph ' + "-"*20 + "\n")
+        graphs = self.model2graphs(model)
+        self.graphs[model.number] = graphs
+        self.save()
+        for graph in self.graphs[model.number]:
+            print(f"Output saved in {graph.filepath}")
```

Comment:
I would ommit this print altogether. I left it there because you had it there :)

Reply:
Oh, I misunderstood what this does. (Reading code instead of trying it is not really a good idea.) Anyhow, I feel it is still a little bit to verbose. We should probably discuss in a meeting what clingraph should print to stdout/stderr.